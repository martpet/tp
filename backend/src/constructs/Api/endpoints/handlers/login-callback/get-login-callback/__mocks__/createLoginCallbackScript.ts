export const createLoginCallbackScript = vi.fn().mockReturnValue({
  script: 'dummyScript',
  cspHash: 'dummyCspHash',
});
