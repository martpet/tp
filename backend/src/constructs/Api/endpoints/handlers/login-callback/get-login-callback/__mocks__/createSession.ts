export const createSession = vi.fn().mockResolvedValue('dummySessionCookie');
