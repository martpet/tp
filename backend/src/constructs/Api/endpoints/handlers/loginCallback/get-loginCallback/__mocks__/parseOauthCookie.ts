import { OauthCookieProps } from '~/constructs/Api/types';

export const dummyOauthCookieProps: OauthCookieProps = {
  stateNonce: 'dummyState',
  idTokenNonce: 'dummyIdTokenNonce',
  codeVerifier: 'dummyCodeVerifier',
};

export const parseOauthCookie = vi.fn().mockReturnValue(dummyOauthCookieProps);
