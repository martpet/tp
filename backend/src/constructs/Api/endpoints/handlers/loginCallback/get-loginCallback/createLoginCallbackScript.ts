import { loginWindowSuccessMessage } from '~/consts';
import { EnvName } from '~/types';
import { createSha256CspHash } from '~/utils';

type CreateLoginCallbackScriptProps = {
  envName: EnvName;
  appDomain: string;
};

export const createLoginCallbackScript = ({
  envName,
  appDomain,
}: CreateLoginCallbackScriptProps) => {
  const targetOrigin = envName === 'personal' ? '*' : `https://${appDomain}`;
  const script = `opener.postMessage("${loginWindowSuccessMessage}", "${targetOrigin}")`;

  return { script, cspHash: createSha256CspHash(script) };
};
