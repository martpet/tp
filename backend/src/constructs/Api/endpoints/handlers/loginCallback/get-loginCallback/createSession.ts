import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import cookie from 'cookie';
import crypto from 'crypto';
import millis from 'milliseconds';

import { OauthTokens } from '~/constructs/Api/types';
import { cookieName } from '~/constructs/Api/utils';
import { refreshTokenValidityInDays, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

type CreateSessionProps = {
  tokens: OauthTokens;
};

export const createSession = async ({ tokens }: CreateSessionProps) => {
  const { idTokenPayload, refreshToken, idToken } = tokens;
  const { envName } = globalLambdaProps;
  const sessionId = crypto.randomUUID();
  const created = Date.now();
  const refreshTokenExpires = created + millis.days(refreshTokenValidityInDays);

  const putCommandItem: SessionsTableItem = {
    id: sessionId,
    userId: idTokenPayload.sub,
    created,
    refreshToken,
    refreshTokenExpires,
    idToken,
  };

  const putCommand = new PutCommand({
    TableName: sessionsTableOptions.tableName,
    Item: putCommandItem,
  });

  await ddbDocClient.send(putCommand);

  const sessionCookie = cookie.serialize(cookieName('sessionId'), sessionId, {
    secure: true,
    httpOnly: true,
    sameSite: envName !== 'personal' || 'none',
    expires: new Date(refreshTokenExpires),
  });

  return sessionCookie;
};
