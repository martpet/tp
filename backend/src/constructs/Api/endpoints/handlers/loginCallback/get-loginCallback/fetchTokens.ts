import camelcaseKeys from 'camelcase-keys';
import fetch from 'node-fetch';

import { FetchTokensResponse, OauthTokens } from '~/constructs/Api/types';
import { getIdTokenPayload } from '~/constructs/Api/utils';
import { authPaths } from '~/consts';

type FetchTokensProps = {
  code: string;
  codeVerifier: string;
  clientId: string;
  authDomain: string;
  loginCallbackUrl: string;
};

export const fetchTokens = async ({
  code,
  codeVerifier,
  clientId,
  authDomain,
  loginCallbackUrl,
}: FetchTokensProps): Promise<OauthTokens> => {
  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: clientId,
      code,
      code_verifier: codeVerifier,
      redirect_uri: loginCallbackUrl,
    }),
  });

  const data = (await response.json()) as FetchTokensResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  return {
    ...camelcaseKeys(data),
    idTokenPayload: getIdTokenPayload(data.id_token),
  };
};
