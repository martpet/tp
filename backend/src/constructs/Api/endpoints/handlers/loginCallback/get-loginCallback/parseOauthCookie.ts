import {
  APIGatewayProxyEventV2,
  OauthCookieProps,
  parseEventCookies,
} from 'lambda-layer';

export const parseOauthCookie = (event: APIGatewayProxyEventV2) => {
  try {
    const { oauth } = parseEventCookies<'/loginCallback'>(event);
    if (!oauth) throw new Error();
    return JSON.parse(oauth) as Partial<OauthCookieProps>;
  } catch (e) {
    console.error(e);
    return {};
  }
};
