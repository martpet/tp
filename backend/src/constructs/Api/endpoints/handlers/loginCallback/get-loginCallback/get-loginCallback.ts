import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import { StatusCodes } from 'http-status-codes';

import { OauthTokens, ProcessEnv, QueryStringParameters } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils';
import { appEnvs, loginPopupSuccessMessage } from '~/consts';

import { createSession } from './createSession';
import { fetchTokens } from './fetchTokens';
import { parseOauthCookie } from './parseOauthCookie';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { cdkEnv } = globalLambda;

  const { state, code } = Object(
    event.queryStringParameters
  ) as QueryStringParameters<'/loginCallback'>;

  const { clientId, authDomain, loginCallbackUrl, envName } =
    process.env as ProcessEnv<'/loginCallback'>;

  const { stateNonce, idTokenNonce, codeVerifier } = parseOauthCookie(event);

  if (!state || !code) {
    return errorResponse('1cByWITGGHw', { statusCode: StatusCodes.BAD_REQUEST });
  }

  if (!clientId || !authDomain || !loginCallbackUrl || !envName) {
    return errorResponse('XuJRO7_43O');
  }

  if (!stateNonce || !idTokenNonce || !codeVerifier) {
    return errorResponse('OSUiwUI_DG');
  }

  if (state !== stateNonce) {
    return errorResponse('eWe87M5edc', { statusCode: StatusCodes.UNAUTHORIZED });
  }

  let tokens: OauthTokens;
  let sessionCookie: string;

  try {
    tokens = await fetchTokens({
      code,
      codeVerifier,
      clientId,
      authDomain,
      loginCallbackUrl,
    });
  } catch (error) {
    return errorResponse('-glnzeksvh', { error });
  }

  if (tokens.idTokenPayload.nonce !== idTokenNonce) {
    return errorResponse('C25fjaal7o', { statusCode: StatusCodes.UNAUTHORIZED });
  }

  try {
    sessionCookie = await createSession({ tokens, envName });
  } catch (error) {
    return errorResponse('h7UMnEUDvF', { error });
  }

  const postMessageTargetOrigin =
    cdkEnv === 'personal' ? '*' : `https://${appEnvs[cdkEnv].appDomain}`;

  return {
    statusCode: StatusCodes.OK,
    cookies: [sessionCookie],
    headers: { 'Content-Type': 'text/html' },
    body: `<script>opener.postMessage("${loginPopupSuccessMessage}", "${postMessageTargetOrigin}")</script>`,
  };
};
