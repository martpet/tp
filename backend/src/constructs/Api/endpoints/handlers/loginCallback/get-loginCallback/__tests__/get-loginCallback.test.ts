import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { OauthCookieProps } from '~/constructs/Api/types';
import { itResolvesWithErrorResponse } from '~/constructs/Api/utils';

import { dummyOauthTokens } from '../__mocks__/fetchTokens';
import { dummyOauthCookieProps } from '../__mocks__/parseOauthCookie';
import { createLoginCallbackScript } from '../createLoginCallbackScript';
import { createSession } from '../createSession';
import { fetchTokens } from '../fetchTokens';
import { handler } from '../get-loginCallback';
import { parseOauthCookie } from '../parseOauthCookie';

vi.importMock('../createSession');

vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('../fetchTokens');
vi.mock('../parseOauthCookie');
vi.mock('../createSession');
vi.mock('../createLoginCallbackScript');

const args = [
  {
    cookies: 'dummyCookies',
    queryStringParameters: {
      code: 'dummyCode',
      state: 'dummyState',
    },
  },
] as unknown as Parameters<APIGatewayProxyHandlerV2>;

beforeEach(() => {
  process.env.authDomain = 'dummyAuthDomain';
  process.env.clientId = 'dummyClientId';
  process.env.loginCallbackUrl = 'dummyLoginCallbackUrl';
  process.env.envName = 'dummyEnvName';
});

describe('"get-loginCallback" handler', () => {
  it('calls "parseOauthCookie" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(parseOauthCookie).mock.calls).toMatchSnapshot();
  });

  it('calls "fetchTokens" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(fetchTokens).mock.calls).toMatchSnapshot();
  });

  it('calls "createSession" with corrects args', async () => {
    await handler(...args);
    expect(vi.mocked(createSession).mock.calls).toMatchSnapshot();
  });

  it('calls "createLoginCallbackScript" with corrects args', async () => {
    await handler(...args);
    expect(vi.mocked(createLoginCallbackScript).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  describe.each(['code', 'state'])(
    'when "%s" query string parameter is missing',
    (key) => {
      const argsClone = structuredClone(args);
      const { queryStringParameters } = argsClone[0] as Required<typeof argsClone[0]>;
      delete queryStringParameters[key];
      itResolvesWithErrorResponse(handler, argsClone);
    }
  );

  describe.each(['clientId', 'authDomain', 'loginCallbackUrl'])(
    'when "%s" env var is missing',
    (key) => {
      beforeEach(() => {
        delete process.env[key];
      });
      itResolvesWithErrorResponse(handler, args);
    }
  );

  describe.each(['stateNonce', 'idTokenNonce', 'codeVerifier'])(
    'when "oauth" cookie content is missing "%s" key',
    (key) => {
      const oauthCookiePropsClone = structuredClone(dummyOauthCookieProps);
      delete oauthCookiePropsClone[key as keyof OauthCookieProps];
      beforeEach(() => {
        vi.mocked(parseOauthCookie).mockReturnValueOnce(oauthCookiePropsClone);
      });
      itResolvesWithErrorResponse(handler, args);
    }
  );

  describe('when "fetchTokens" rejects', () => {
    beforeEach(() => {
      vi.mocked(fetchTokens).mockRejectedValueOnce(new Error('dummyErrorMessage'));
    });
    itResolvesWithErrorResponse(handler, args);
  });

  describe('when "createSession" rejects', () => {
    beforeEach(() => {
      vi.mocked(createSession).mockRejectedValueOnce(new Error('dummyErrorMessage'));
    });
    itResolvesWithErrorResponse(handler, args);
  });

  describe('when "stateNonce" from "oauth" cookie does not match with "state" from query strings params', () => {
    const argsClone = structuredClone(args);
    const { queryStringParameters } = argsClone[0] as Required<typeof argsClone[0]>;
    queryStringParameters.state = 'differentState';
    itResolvesWithErrorResponse(handler, argsClone);
  });

  describe('when "idTokenNonce" from "oauth" cookie does not match with "nonce" from the "idToken" payload', () => {
    const tokensPropsClone = structuredClone(dummyOauthTokens);
    tokensPropsClone.idTokenPayload.nonce = 'differentNonce';
    beforeEach(() => {
      vi.mocked(fetchTokens).mockResolvedValueOnce(tokensPropsClone);
    });
    itResolvesWithErrorResponse(handler, args);
  });
});
