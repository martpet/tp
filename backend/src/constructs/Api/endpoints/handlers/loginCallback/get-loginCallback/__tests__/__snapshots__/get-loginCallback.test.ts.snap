// Vitest Snapshot v1

exports[`"get-loginCallback" handler > calls "createSession" with corrects args 1`] = `
[
  [
    {
      "envName": "dummyEnvName",
      "tokens": {
        "accessToken": "dummyAccessToken",
        "expiresIn": 123,
        "idToken": "dummyIdToken",
        "idTokenPayload": {
          "aud": "dummyAud",
          "nonce": "dummyIdTokenNonce",
          "sub": "dummySub",
        },
        "refreshToken": "dummyRefreshToken",
      },
    },
  ],
]
`;

exports[`"get-loginCallback" handler > calls "fetchTokens" with correct args 1`] = `
[
  [
    {
      "authDomain": "dummyAuthDomain",
      "clientId": "dummyClientId",
      "code": "dummyCode",
      "codeVerifier": "dummyCodeVerifier",
      "loginCallbackUrl": "dummyLoginCallbackUrl",
    },
  ],
]
`;

exports[`"get-loginCallback" handler > calls "parseOauthCookie" with correct args 1`] = `
[
  [
    {
      "cookies": "dummyCookies",
      "queryStringParameters": {
        "code": "dummyCode",
        "state": "dummyState",
      },
    },
  ],
]
`;

exports[`"get-loginCallback" handler > resolves with a correct value 1`] = `
{
  "body": "<script>opener.postMessage(\\"login-success\\", \\"https://trip.pictures\\")</script>",
  "cookies": [
    "dummySessionCookie",
  ],
  "headers": {
    "Content-Type": "text/html",
  },
  "statusCode": 200,
}
`;

exports[`"get-loginCallback" handler > when "authDomain" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-loginCallback" handler > when "authDomain" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "clientId" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-loginCallback" handler > when "clientId" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "code" query string parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1cByWITGGHw",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`"get-loginCallback" handler > when "code" query string parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "createSession" rejects > calls "errorResponse" with correct args 1`] = `
[
  [
    "h7UMnEUDvF",
    {
      "error": [Error: dummyErrorMessage],
    },
  ],
]
`;

exports[`"get-loginCallback" handler > when "createSession" rejects > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "envName" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-loginCallback" handler > when "envName" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "fetchTokens" rejects > calls "errorResponse" with correct args 1`] = `
[
  [
    "-glnzeksvh",
    {
      "error": [Error: dummyErrorMessage],
    },
  ],
]
`;

exports[`"get-loginCallback" handler > when "fetchTokens" rejects > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "globalLambda.cdkEnv" is "personal" > resolves with a correct value 1`] = `
{
  "body": "<script>opener.postMessage(\\"login-success\\", \\"*\\")</script>",
  "cookies": [
    "dummySessionCookie",
  ],
  "headers": {
    "Content-Type": "text/html",
  },
  "statusCode": 200,
}
`;

exports[`"get-loginCallback" handler > when "idTokenNonce" from "oauth" cookie does not match with "nonce" from the "idToken" payload > calls "errorResponse" with correct args 1`] = `
[
  [
    "C25fjaal7o",
    {
      "statusCode": 401,
    },
  ],
]
`;

exports[`"get-loginCallback" handler > when "idTokenNonce" from "oauth" cookie does not match with "nonce" from the "idToken" payload > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "loginCallbackUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-loginCallback" handler > when "loginCallbackUrl" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "oauth" cookie content is missing "codeVerifier" key > calls "errorResponse" with correct args 1`] = `
[
  [
    "OSUiwUI_DG",
  ],
]
`;

exports[`"get-loginCallback" handler > when "oauth" cookie content is missing "codeVerifier" key > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "oauth" cookie content is missing "idTokenNonce" key > calls "errorResponse" with correct args 1`] = `
[
  [
    "OSUiwUI_DG",
  ],
]
`;

exports[`"get-loginCallback" handler > when "oauth" cookie content is missing "idTokenNonce" key > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "oauth" cookie content is missing "stateNonce" key > calls "errorResponse" with correct args 1`] = `
[
  [
    "OSUiwUI_DG",
  ],
]
`;

exports[`"get-loginCallback" handler > when "oauth" cookie content is missing "stateNonce" key > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "state" query string parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1cByWITGGHw",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`"get-loginCallback" handler > when "state" query string parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-loginCallback" handler > when "stateNonce" from "oauth" cookie does not match with "state" from query strings params > calls "errorResponse" with correct args 1`] = `
[
  [
    "eWe87M5edc",
    {
      "statusCode": 401,
    },
  ],
]
`;

exports[`"get-loginCallback" handler > when "stateNonce" from "oauth" cookie does not match with "state" from query strings params > resolves with a correct value 1`] = `"dummyErrorResponse"`;
